neps/optimizers/models/deepGP.py

```diff
@@ -0,0 +1,572 @@
+
+    # Created from the above arguments
+    # TODO: Lift this out of DeepGP and let the optimizer worry about pre-processing
+    preprocessor: DeepGPDataTransformer = field(init=False)
+
+    # Post fit parameters, following scikit-learn convention of appending an underscore
+    model_: GPRegressionModel | None = field(init=False)
+    likelihood_: gpytorch.likelihoods.GaussianLikelihood | None = field(init=False)
+    nn_: NeuralFeatureExtractor | None = field(init=False)
+    projected_x_train_: torch.Tensor | None = field(init=False)
+    y_train_: torch.Tensor | None = field(init=False)
+    y_bounds_: tuple[float, float] | None = field(init=False)
+
+    def __post_init__(self):
+        if any(isinstance(h, GraphParameter) for h in self.pipeline_space.values()):
+            raise ValueError("Graph parameters are not supported for DeepGP")
+
+        if self.normalize_budget:
+            budget_bounds = (pipeline_space.fidelity.lower, pipeline_space.fidelity.upper)  # type: ignore
+        else:
+            budget_bounds = None
+
+        if self.checkpointing:
+            assert (
+                self.root_directory is not None
+            ), "neps root_directory must be provided for the checkpointing"
+            self.checkpoint_path = self.root_directory / self.checkpoint_file
+
+        self.preprocessor = DeepGPDataTransformer(
+            space=self.pipeline_space,
+            fidelity_bounds=budget_bounds,
+            normalize_y=self.normalize_y,
+            min_learning_curve_length=self.neural_network_args.get("cnn_kernel_size", 3),
+            learning_curve_pad_value=self.learning_curve_pad_value,
+            device=self.device,
+        )
+        self.model_ = None
+        self.likelihood_ = None
+        self.nn_ = None
+
+    def fit(
+        self,
+        x_train: list[SearchSpace],
+        y_train: list[float],
+        learning_curves: list[list[float]],
+    ):
+        x_, train_budget = self.preprocessor.encode_configs(x_train)
+        curves = self.preprocessor.encode_learning_curves(learning_curves)
+        y_, y_bounds = self.preprocessor.encode_y(y_train)
+
+        # Required for predictions later
+        self.y_train_ = y_
+        self.y_bounds_ = y_bounds
+
+        input_dim = x_.shape[1]
+
+        # Initial state
+        likelihood = gpytorch.likelihoods.GaussianLikelihood().to(self.device)
+        model = GPRegressionModel(train_x=x_, train_y=y_, likelihood=likelihood).to(
+            self.device
+        )
+        nn = NeuralFeatureExtractor(input_dim, **self.neural_network_args).to(self.device)
+
+        # If checkpointing and we are improving, load existing model
+        if self.checkpointing and self.checkpoint_path.exists():
+            assert self.root_directory is not None
+
+            non_improvement_steps = count_non_improvement_steps(self.root_directory)
+            if non_improvement_steps < self.perf_patience:
```

Comment:
The last change on the other branch is the [refine condition](https://github.com/automl/neps/blob/d6e123faf41dde2461241a20b2014e4345d9a2a3/neps/optimizers/bayesian_optimization/models/deepGP.py#L268)